x509/src/Cardano/X509/Configuration.hs

```diff
@@ -196,7 +202,74 @@ fromConfiguration TLSConfiguration{..} DirConfiguration{..} certGenKeys (caPub,
 
 
 --
--- INTERNALS / UTILS
+-- Effectful Functions
+--
+
+-- | Type-alias for signature readability
+newtype ConfigurationKey = ConfigurationKey
+    { getConfigurationKey :: String
+    } deriving (Eq, Show)
+
+-- | Decode a configuration file (.yaml). The expected file structure is:
+--     <configuration-key>:
+--       tls:
+--         ca: ...
+--         server: ...
+--         clients: ...
+--
+-- where the 'configuration-key' represents the target environment (dev, test,
+-- bench, etc.).
+decodeConfigFile
+    :: (MonadIO m, MonadFail m)
+    => ConfigurationKey -- ^ Target configuration Key
+    -> FilePath         -- ^ Target configuration file
+    -> m TLSConfiguration
+decodeConfigFile (ConfigurationKey cKey) filepath =
+    decodeFileMonad filepath >>= parseMonad parser
+  where
+    errMsg key = "Invalid TLS Configuration: property '"<> key <> "' " <>
+        "not found in configuration file."
+
+    decodeFileMonad = (liftIO . decodeFileEither) >=> either (fail . show) return
+
+    parser = withObject "TLS Configuration" (parseK cKey >=> parseK "tls")
+
+    parseK :: FromJSON a => String -> Aeson.Object -> Aeson.Parser a
+    parseK key = maybe (fail $ errMsg key) parseJSON . HM.lookup (toText key)
+
+
+-- | Generate & sign a certificate from a certificate description
+genCertificate
+    :: CertDescription IO PublicKey PrivateKey filename
+    -> IO (PrivateKey, SignedCertificate)
+genCertificate desc = do
+    ((pub, priv), now) <- (,) <$> (certGenKeys desc) <*> dateCurrent
+
+    let conf = certConfiguration desc
+    let cert = X509.Certificate
+            { X509.certVersion      = 2
+            , X509.certSerial       = fromIntegral (certSerial desc)
+            , X509.certSignatureAlg = signAlgRSA256
+            , X509.certValidity     = (addMinutes (-1) now, addDays (certExpiryDays conf) now)
+            , X509.certPubKey       = PubKeyRSA pub
+            , X509.certExtensions   = Extensions (Just $ certExtensions desc)
+            , X509.certIssuerDN     = certIssuer desc
+            , X509.certSubjectDN    = certSubject desc
+            }
+
+    (priv,) <$> signCertificate (certSigningKey desc) cert
+  where
+    addDays :: Int -> DateTime -> DateTime
+    addDays n time@(DateTime date _) =
+        time { dtDate = dateAddPeriod date (mempty { periodDays = n }) }
+
+    addMinutes :: Int -> DateTime -> DateTime
+    addMinutes n time =
+        timeAdd time (Minutes $ fromIntegral n)
```

Comment:
`genCertificate` used to be defined in `gencerts/Main.hs`, I pulled it into the library since it's quite a re-usable piece.